### DISCUSSION TITLE
Sentinel2.mask_clouds_and_shadows add optional argument fill value

### DISCUSSION THREAD

**Discussion by orianif** [EOA-team/eodal#63] [Announcements]
It would be great if the method Sentinel2.mask_clouds_and_shadows would have an optional argument to specify the filling value, which for now is 0 by default. In case raw pixel values are extracted and treated programmatically one may want to label missing values differently (e.g. as nans).

Best,
Fabio

**Comment by orianif** [comment]
Hi Lukas,

what about a datatype check on the user entry?

Fabio Oriani

Researcher - Agroscope, Mountain Grassland Team

Earth Observation for Agroecosystems (EOA) Team

https://www.eoa-team.nethttps://www.eoa-team.net/profile/FabioOriani/

Associate Editor - Computers & Geosciences
https://www.journals.elsevier.com/computers-and-geosciences/

Inviato da Proton Mail mobile

-------- Messaggio originale --------
Il 5 mag 2023, 20:14, Lukas Valentin Graf ha scritto:

**Comment by lukasValentin** [comment]
Hi @orianif 

thanks for the suggestions. While I think we can discuss about fill values, I have two comments:

First: The mask method actually should not alter data values. It uses `np.ma.MaskedArray` to mask out pixels by setting their entry in the `mask` part of the `MaskedArray` to False. Operations in EOdal always check for the `is_masked_array` attribute to handle these arrays correctly. So, it should not happen that *data* values are set to zero (which is the no-data value of Sentinel-2 defined by ESA by the way). In case it happens, could you please post a code snippet for reproducing the issue?

Second, I think a major issue with using custom fill values is the data type. 'nan' will produce an error when working with integer values, which are widely used for remote sensing data products (e.g., 16bit integers for Sentinel-2). So, the question would be how to handle such "wrong" fill values?

Best,
Lukas

**Comment by lukasValentin** [comment]
Hi @orianif, is this issue still applicable? If so, can you please post a minimum reproducible example so that I can see what's going on in your case?

**Comment by lukasValentin** [comment]
Since we now fixed several bugs related to the numpy data type and fill value of masked array I'd close this discussion for the time being.
